aptos-move/mvhashmap/src/versioned_code.rs

```diff
@@ -0,0 +1,205 @@
+
+impl<V: TransactionWrite, X: Executable> Default for VersionedValue<V, X> {
+    fn default() -> Self {
+        VersionedValue::new()
+    }
+}
+
+impl<K: Hash + Clone + Eq, V: TransactionWrite, X: Executable> VersionedCode<K, V, X> {
+    pub(crate) fn new() -> Self {
+        Self {
+            values: DashMap::new(),
+        }
+    }
+
+    pub(crate) fn mark_estimate(&self, key: &K, txn_idx: TxnIndex) {
+        let v = self.values.get(key).expect("Path must exist");
+        v.versioned_map
+            .get(&txn_idx)
+            .expect("Entry by the txn must exist to mark estimate")
+            .mark_estimate();
+    }
+
+    pub(crate) fn delete(&self, key: &K, txn_idx: TxnIndex) {
+        // TODO: investigate logical deletion.
+        let mut v = self.values.get_mut(key).expect("Path must exist");
+        v.versioned_map.remove(&txn_idx);
+    }
+
+    fn read(&self, key: &K, txn_idx: TxnIndex) -> anyhow::Result<(Arc<V>, HashValue), MVCodeError> {
+        use MVCodeError::*;
+
+        match self.values.get(key) {
+            Some(v) => {
+                if let Some((idx, entry)) = v.versioned_map.range(0..txn_idx).next_back() {
+                    let flag = entry.flag();
+                    if flag == FLAG_ESTIMATE {
+                        // Found a dependency.
+                        return Err(Dependency(*idx));
+                    }
+                    // The entry should be populated.
+                    debug_assert!(flag == FLAG_DONE);
+
+                    Ok((entry.module.clone(), entry.hash))
+                } else {
+                    Err(NotFound)
+                }
+            },
+            None => Err(NotFound),
+        }
+    }
+
+    pub(crate) fn write(&self, key: &K, txn_idx: TxnIndex, data: V) {
+        let mut v = self.values.entry(key.clone()).or_default();
+        v.versioned_map.insert(
+            txn_idx,
+            CachePadded::new(Entry::new_write_from(FLAG_DONE, data)),
+        );
+    }
+
+    pub(crate) fn store_executable(
+        &self,
+        key: &K,
+        descriptor: ExecutableDescriptor,
+        executable: X,
+    ) {
+        let x = Arc::new(executable);
+        match descriptor {
+            ExecutableDescriptor::Published(hash) => {
+                let mut v = self.values.get_mut(key).expect("Path must exist");
+                v.executables.entry(hash).or_insert(x);
+            },
+            ExecutableDescriptor::Storage => {
+                let mut v = self.values.entry(key.clone()).or_default();
+                v.base_executable.get_or_insert(x);
+            },
+        };
+    }
+
+    pub(crate) fn fetch_code(
+        &self,
+        key: &K,
+        txn_idx: TxnIndex,
+    ) -> anyhow::Result<MVCodeOutput<V, X>, MVCodeError> {
+        use MVCodeError::*;
+        use MVCodeOutput::*;
+
+        match self.read(key, txn_idx) {
+            Ok((module, hash)) => {
+                let v = self.values.get(key).expect("Path must exist");
```

Comment:
this seems unnecessary if `read` returns the Value directly

Reply:
Moved read to be a method on VersionedValue, looks a lot cleaner now